File: hub/sdk/python-client/src/central_auth_hub/state.py
"""State (CSRF token) — RFC 6749 §10.12."""

from __future__ import annotations

import hmac
import secrets

from .errors import StateError


def generate_state() -> str:
    """32-char hex random state."""
    return secrets.token_hex(16)


def verify_state(expected: str, provided: str) -> None:
    """Constant-time compare — raise StateError on mismatch."""
    if not expected:
        raise StateError("No state to verify (session expired?)")
    if not hmac.compare_digest(expected, provided):
        raise StateError("State mismatch — possible CSRF attack")

File: hub/sdk/node-client/src/state.ts
import { randomBytes, timingSafeEqual } from "node:crypto";
import { StateError } from "./errors.js";

/** Generate 32-char hex random state (16 bytes). */
export function generateState(): string {
  return randomBytes(16).toString("hex");
}

/** Constant-time compare — throws StateError on mismatch. */
export function verifyState(expected: string, provided: string): void {
  if (!expected) throw new StateError("No state to verify (session expired?)");
  const a = Buffer.from(expected);
  const b = Buffer.from(provided);
  if (a.length !== b.length || !timingSafeEqual(a, b)) {
    throw new StateError("State mismatch — possible CSRF attack");
  }
}
